pkg/image/controller/signature/signature_import_controller.go

```diff
@@ -0,0 +1,185 @@
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"time"
+
+	"github.com/containers/image/docker"
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/controller"
+
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	informers "github.com/openshift/origin/pkg/image/generated/informers/internalversion/image/internalversion"
+	imageclient "github.com/openshift/origin/pkg/image/generated/internalclientset"
+)
+
+type SignatureImportController struct {
+	imageClient    imageclient.Interface
+	imageCache     cache.Store
+	imageHasSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	ctx          context.Context
+	fetchTimeout time.Duration
+}
+
+func NewSignatureImportController(ctx context.Context, imageClient imageclient.Interface, imageInformer informers.ImageInformer, resyncInterval, fetchTimeout time.Duration) *SignatureImportController {
+	controller := &SignatureImportController{
+		queue:          workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ctx:            ctx,
+		fetchTimeout:   fetchTimeout,
+		imageClient:    imageClient,
+		imageCache:     imageInformer.Informer().GetStore(),
+		imageHasSynced: imageInformer.Informer().HasSynced,
+	}
+
+	imageInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			image := obj.(*imageapi.Image)
+			glog.V(4).Infof("Adding image %s", image.Name)
+			controller.enqueueImage(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			image := cur.(*imageapi.Image)
+			glog.V(4).Infof("Updating image %s", image.Name)
+			controller.enqueueImage(cur)
+		},
+	}, resyncInterval)
+
+	return controller
+}
+
+func (s *SignatureImportController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer s.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(stopCh, s.imageHasSynced) {
+		return
+	}
+
+	glog.V(5).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+
+}
+
+func (s *SignatureImportController) worker() {
+	for {
+		if !s.work() {
+			return
+		}
+	}
+}
+
+// work returns true if the worker thread should continue
+func (s *SignatureImportController) work() bool {
+	key, quit := s.queue.Get()
+	if quit {
+		return false
+	}
+	defer s.queue.Done(key)
+
+	if err := s.syncImageSignatures(key.(string)); err == nil {
+		// this means the request was successfully handled.  We should "forget" the item so that any retry
+		// later on is reset
+		s.queue.Forget(key)
+	} else {
+		// if we had an error it means that we didn't handle it, which means that we want to requeue the work
+		utilruntime.HandleError(fmt.Errorf("error syncing service, it will be retried: %v", err))
```

Comment:
not a service

Reply:
dammit... now you know from where I copied this ;-)